### ISSUE TITLE
Ensure performance and caching for authentication token

### ISSUE THREAD

**Issue by jonesbusy** [oras-project/oras-java#17]

**Comment by sparsh3007** [comment]
No, it's not there in Go SDK. I see the logic for expiry of token is already implemented. 

My bad, I will follow the Go SDK to implement caching for authentication.

**Comment by jonesbusy** [comment]
I think the token refresh logic is fragile right now. Specially when ORAS is used as a library and not a program (like a CLI). A single registry can be used to access different repositories, with different scopes

- Get inspiration to https://github.com/oras-project/oras-go/blob/f7a6126986d97f498be7b516ebf5478a49a6df86/registry/remote/auth/client.go#L345
- Move refreshToken to HTTP client instead of BearerTokenAuth (that should just store the current token)
- Store and cache token per scope, service etc
- Don't change the state of Authentication in Registry (remove updateAuthentication). Authentication should be passed each time to the client

Typically some issue

- Registry get library/alpine (with or without basic auth)
- Request Bearer token
- Get scope for library/alpine:pull
- Get library/alpine with Bearer token (switch auth)
- Get library/alpine
- Server return 401, wrong scope (error="insufficient_scope")
- Request Bearer token (but passing previous token, not compliant. Docker.io return 500)

Due to switch auth, the initial credentials are lost

We need absolutely to remove the state of auth on the Registry and implement correctly refresh token

Caching can be a second step, when we get 401 we just request a new token with new scope
